mm2src/mm2_metrics/src/native.rs

```diff
@@ -0,0 +1,491 @@
+            let counter = recorder.register_counter(&key);
+            counter.increment($value);
+        };
+    }};
+}
+
+/// Update gauge if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_gauge {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+        use $crate::metrics::Recorder;
+        if let Some(recorder) = $crate::TryRecorder::try_recorder(&$metrics){
+            let key = $crate::native::key_from_str($name);
+            let gauge = recorder.register_gauge(&key);
+            gauge.increment($value);
+        }
+    }};
+
+    // Register and increment gauge with label
+    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => {{
+        use $crate::metrics::Recorder;
+        use $crate::native::from_slice_to_labels;
+        if let Some(recorder) = $crate::TryRecorder::try_recorder(&$metrics){
+            let key = $crate::metrics::Key::from_parts($name, from_slice_to_labels(vec![$($label_key.to_owned(), $label_val.to_owned()),+]));
+            let gauge = recorder.register_gauge(&key);
+            gauge.increment($value);
+        }
+    }};
+}
+
+/// Update gauge if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_timing {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+        use $crate::metrics::Recorder;
+        use key_from_str;
+        if let Some(recorder) = $crate::TryRecorder::try_recorder(&$metrics){
+            let key = key_from_str($name);
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+        }
+    }};
+
+    // Register and record histogram with label
+    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => {{
+        use $crate::metrics::Recorder;
+        use $crate::native::from_slice_to_labels;
+        if let Some(recorder) = $crate::TryRecorder::try_recorder(&$metrics){
+            let key = $crate::metrics::Key::from_parts($name, from_slice_to_labels(vec![$($label_key.to_owned(), $label_val.to_owned()),+]));
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+        }
+    }};
+}
+
+/// Convert a string to metrics Key
+pub fn key_from_str(name: &'static str) -> Key {
+    let key_name = KeyName::from_const_str(name);
+    Key::from_name(key_name)
+}
+
+/// Convert a vector of strings to metric labels
+pub fn from_slice_to_labels(labels: Vec<String>) -> Vec<Label> {
+    let labels = labels.to_vec();
+    let mut new_lab: Vec<_> = vec![];
+    let mut i = 0;
+    for _ in 0..labels.len() - 1 {
+        if i < labels.len() {
+            let label = (labels[i].to_owned(), labels[i + 1].to_owned());
+            new_lab.push(label);
+            i += 2;
+        }
+    }
+
+    new_lab.as_slice().into_labels()
```

Comment:
This code can be simplified by using [Itertools::tuples](https://docs.rs/itertools/0.7.6/itertools/trait.Itertools.html#method.tuples).
For example,
```rust
for (x, y) in vec![1, 2, 3, 4, 5].into_iter().tuples() {
    println!("{} {}", x, y);
}
```
will print
> 1, 2
3, 4

ignoring the last 5th item (as there is no pair for it).
Or you can use [Vec::chunks]():
```rust
for x in vec![1, 2, 3, 4, 5].chunks_exact(2) {
    // Where `x` is a slice `&[i32]`
    println!("{:?}", x);
}
```